### PR TITLE
Ridvan wallet plugin extension

### PR DESCRIPTION
**Pull request by rozaydin**

### CODE REVIEW

**Comment by rozaydin** [comment]
I have completed, automated and manual tests for this PR,
- fixed an account address bug, 
- 3 known tests withing hsd unit test is failing, all others passing.

File: lib/utils/util.js [hunk 1]
```diff
@@ -184,72 +183,13 @@ util.encodeU32 = function encodeU32(num) {
...
- * }}
- */
-
-util.createStrictBatch = function(limit, domainOutputMap) {
-  if (!limit || limit === 0)
-    throw new Error('invalid limit provided!');
+util.createStrictBatch = function (limit, domainOutputMap) {
+  if (!limit || limit === 0) {
+ throw new Error('invalid limit provided!');
+}
```
Comment (RevCBH) [inline lib/utils/util.js:189]: The whitespace is off here. The indenting looks like the function ends here.

File: lib/wallet/rpc.js [hunk 2]
```diff
@@ -2039,41 +2036,6 @@ class RPC extends RPCBase {
     return {name, force, account};
   }
 
-  async sendOpenIdempotent(args, help) {
```
Comment (RevCBH) [inline lib/wallet/rpc.js:2042]: We are completely doing away with these RPC methods, right? Not supporting adding new ones like we are for routes.

File: lib/wallet/wallet.js [hunk 3]
```diff
@@ -101,11 +101,6 @@ class Wallet extends EventEmitter {
       consensus.MAX_BLOCK_OPENS * bidsPerAuction * bidsCacheSizeInBlocks;
     this.sendBidResults = new LRU(bidsCapacity);
 
-    // TODO (bennett) - make this configurable from options
-    const opensCacheSizeInBlocks = 21;
-    const opensCapacity = consensus.MAX_BLOCK_OPENS * opensCacheSizeInBlocks;
-    this.sendOpenResults = new LRU(opensCapacity);
-
     // TODO (bennett) - make this configurable from options
     const updatesCacheSizeInBlocks = 20;
```
Comment (RevCBH) [inline lib/wallet/wallet.js:105]: Why is the updates cache staying around?

**Review by RevCBH — APPROVED** [review]
A whitespace typo and a couple of questions, but looks good

File: lib/wallet/http.js [hunk 4]
```diff
@@ -2164,6 +1751,40 @@ class HTTP extends Server {
...
+   */
+  registerRoute(httpMethod, path, handler, context) {
+    const validHttpMethods = ['post', 'get', 'put', 'patch', 'del'];
+    assert(httpMethod && typeof httpMethod === 'string',
+    'httpMethod parameter must be provided and it must be a string');
+    assert(validHttpMethods.includes(httpMethod.toLowerCase()),
+    'httpMethod must be one of ' + validHttpMethods.join(','));
+
+    // update context
+    context.TransactionOptions = TransactionOptions;
```
Comment (turbomaze) [inline lib/wallet/http.js:1778]: Seems like some of these can be imported directly from the plugin, what is the benefit to passing these in here? Only `this.network` seems like actual state that belongs in the context to me, what do you think

File: lib/wallet/wallet.js [hunk 5]
```diff
@@ -1975,7 +1951,7 @@ class Wallet extends EventEmitter {
       throw new Error(`Bid exceeds lockup value: "${name}".`);
 
     const addr = consecutiveCall ? await this.receiveAddress(acct)
-      : (await this.createReceive()).getAddress();
+      : (await this.createReceive(acct)).getAddress();
```
Comment (turbomaze) [inline lib/wallet/wallet.js:1954]: Nice catch

**Review by turbomaze — CHANGES_REQUESTED** [review]

File: lib/utils/util.js (hunk 1 above)
Comment (rozaydin) [reply lib/utils/util.js:189]: I have moved everything to this PR -> https://github.com/namebasehq/full-node/pull/22
Now the plugin is part of full-node, I have also updated the eslint rulzes (extended from airbnb-base) and applied auto-formatting.

File: lib/wallet/rpc.js (hunk 2 above)
Comment (rozaydin) [reply lib/wallet/rpc.js:2042]: Yes, all interaction between namebase app and wallet will be using http methods from now on. We wont be adding new routes for RPC methods and refactor existing ones to use http ones instead

File: lib/wallet/http.js (hunk 4 above)
Comment (rozaydin) [reply lib/wallet/http.js:1778]: TransactionOptions is declared within wallet http.js and it's not exported. I also agree if we can directly import from the hsd, we should do that way. If there are any i missed please let me know ...

```
class TransactionOptions {
  /**
   * TransactionOptions
   * @alias module:http.TransactionOptions
   * @constructor
   * @param {Validator} valid
   */

  constructor(valid) {
    if (valid)
      return this.fromValidator(valid);
  }

  /**
   * Inject properties from Validator.
   * @private
   * @param {Validator} valid
   * @returns {TransactionOptions}
   */

  fromValidator(valid) {
    assert(valid);

    this.rate = valid.u64('rate');
    this.maxFee = valid.u64('maxFee');
    this.selection = valid.str('selection');
    this.smart = valid.bool('smart');
    this.account = valid.str('account');
    this.locktime = valid.u64('locktime');
    this.sort = valid.bool('sort');
    this.subtractFee = valid.bool('subtractFee');
    this.subtractIndex = valid.i32('subtractIndex');
    this.depth = valid.u32(['confirmations', 'depth']);
    this.paths = valid.bool('paths');
    this.outputs = [];

    if (valid.has('outputs')) {
      const outputs = valid.array('outputs');

      for (const output of outputs) {
        const valid = new Validator(output);

        let addr = valid.str('address');

        if (addr)
          addr = Address.fromString(addr, this.network);

        let covenant = valid.obj('covenant');

        if (covenant)
          covenant = Covenant.fromJSON(covenant);

        this.outputs.push({
          value: valid.u64('value'),
          address: addr,
          covenant: covenant
        });
      }
    }

    return this;
  }

  /*
   * Instantiate transaction options
   * from Validator.
   * @param {Validator} valid
   * @returns {TransactionOptions}
   */

  static fromValidator(valid) {
    return new this().fromValidator(valid);
  }
}
```

File: lib/wallet/wallet.js (hunk 3 above)
Comment (rozaydin) [reply lib/wallet/wallet.js:105]: The caching logic is staying the wallet.js file still, i just removed opencache, but rest is still in hsd code. The plugin is sort of providing the http handler methods (within wallet http.js) at the moment and it's not providing a way to extend wallet.js yet.
```

    // TODO (bennett) - make this configurable from options
    const bidsCapacity =
      consensus.MAX_BLOCK_OPENS * bidsPerAuction * bidsCacheSizeInBlocks;
    this.sendBidResults = new LRU(bidsCapacity);

    // TODO (bennett) - make this configurable from options
    const updatesCacheSizeInBlocks = 20;
    const updatesPerAuction = 1; // average in March 2020 is ~3.1
    const updateCapacity =
      consensus.MAX_BLOCK_OPENS * updatesPerAuction * updatesCacheSizeInBlocks;
    this.sendUpdateResults = new LRU(updateCapacity);

    this.hsdTransferResults = new LRU(25000);
    this.sendTransferResults = new LRU(25000);
    this.sendFinalizeResults = new LRU(25000);
    this.sendFinishResults = new LRU(25000);
    this.sendRevealResults = new LRU(10000);
```

**Pull request closed without merging**